fs/vfs/fs_dup.c

```diff
@@ -50,6 +50,47 @@
  * Public Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: file_dup
+ *
+ * Description:
+ *   Equivalent to the standard dup() function except that it
+ *   accepts a struct file instance instead of a file descriptor.
+ *
+ * Returned Value:
+ *   The new file descriptor is returned on success; a negated errno value
+ *   is returned on any failure.
+ *
+ ****************************************************************************/
+
+int file_dup(FAR struct file *filep, int minfd)
+{
+  struct file filep2;
+  int fd2;
+  int ret;
+
+  /* Let file_dup2() do the real work */
+
+  memset(&filep2, 0, sizeof(filep2));
+  ret = file_dup2(filep, &filep2);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  /* Then allocate a new file descriptor for the inode */
+
+  fd2 = files_allocate(filep2.f_inode, filep2.f_oflags,
+                       filep2.f_pos, filep2.f_priv, minfd);
+  if (fd2 < 0)
+    {
+      file_close(&filep2);
+      return fd2;
```

Comment:
Should we be concerned if f`iles_allocate` fails at `_files_semtake` and returns other values than `EBADF` and `EMFILE`?

Reply:
It should be fine to return -EINTR not -EBADF/-EMFILE since it report the real problem.